build/warp.py

```diff
@@ -0,0 +1,89 @@
+
+def quad_to_rect(quad):
+    assert(len(quad) == 8)
+    assert(quad_as_rect(quad))
+    return (quad[0], quad[1], quad[4], quad[3])
+
+def rect_to_quad(rect):
+    assert(len(rect) == 4)
+    return (rect[0], rect[1], rect[0], rect[3], rect[2], rect[3], rect[2], rect[1])
+
+def shape_to_rect(shape):
+    assert(len(shape) == 2)
+    return (0, 0, shape[0], shape[1])
+
+def griddify(rect, w_div, h_div):
+    w = rect[2] - rect[0]
+    h = rect[3] - rect[1]
+    x_step = w / float(w_div)
+    y_step = h / float(h_div)
+    y = rect[1]
+    grid_vertex_matrix = []
+    for _ in range(h_div + 1):
+        grid_vertex_matrix.append([])
+        x = rect[0]
+        for _ in range(w_div + 1):
+            grid_vertex_matrix[-1].append([int(x), int(y)])
+            x += x_step
+        y += y_step
+    grid = np.array(grid_vertex_matrix)
+    return grid
+
+def distort_grid(org_grid, max_shift):
+    new_grid = np.copy(org_grid)
+    x_min = np.min(new_grid[:, :, 0])
+    y_min = np.min(new_grid[:, :, 1])
+    x_max = np.max(new_grid[:, :, 0])
+    y_max = np.max(new_grid[:, :, 1])
+    new_grid += np.random.randint(- max_shift, max_shift + 1, new_grid.shape)
+    new_grid[:, :, 0] = np.maximum(x_min, new_grid[:, :, 0])
+    new_grid[:, :, 1] = np.maximum(y_min, new_grid[:, :, 1])
+    new_grid[:, :, 0] = np.minimum(x_max, new_grid[:, :, 0])
+    new_grid[:, :, 1] = np.minimum(y_max, new_grid[:, :, 1])
+    return new_grid
+
+def grid_to_mesh(src_grid, dst_grid):
+    assert(src_grid.shape == dst_grid.shape)
+    mesh = []
+    for i in range(src_grid.shape[0] - 1):
+        for j in range(src_grid.shape[1] - 1):
+            src_quad = [src_grid[i    , j    , 0], src_grid[i    , j    , 1],
+                        src_grid[i + 1, j    , 0], src_grid[i + 1, j    , 1],
+                        src_grid[i + 1, j + 1, 0], src_grid[i + 1, j + 1, 1],
+                        src_grid[i    , j + 1, 0], src_grid[i    , j + 1, 1]]
+            dst_quad = [dst_grid[i    , j    , 0], dst_grid[i    , j    , 1],
+                        dst_grid[i + 1, j    , 0], dst_grid[i + 1, j    , 1],
+                        dst_grid[i + 1, j + 1, 0], dst_grid[i + 1, j + 1, 1],
+                        dst_grid[i    , j + 1, 0], dst_grid[i    , j + 1, 1]]
+            dst_rect = quad_to_rect(dst_quad)
+            mesh.append([dst_rect, src_quad])
+    return mesh
+
+def main():
+    directory = os.fsencode("./preprocessed")
+    count = 0
+    for file in os.listdir(directory):
+     filename = os.fsdecode(file)
+     to_open = "./preprocessed/" + filename
+     if filename.endswith(".png"):
+        im = Image.open(to_open)
+        dst_grid = griddify(shape_to_rect(im.size), 4, 4)
+        src_grid = distort_grid(dst_grid, 50)
+        mesh = grid_to_mesh(src_grid, dst_grid)
+        im = im.transform(im.size, Image.MESH, mesh)
+        im.save("./processed/" + "IMG_" + str(count) + ".png")
+        count += 1
+
+if __name__ == "__main__":
+    main()
+
```

Comment:
Could you please include a few example images before and after distortion?

Reply:
I signed an NDA with docucharm, so I can't show the sensitive content, but I have a blurred set of images of before and after warp that I'll send through email

Reply:
https://stackoverflow.com/questions/5071063/is-there-a-library-for-image-warping-image-morphing-for-python-with-controlled

Reply:
I actually used that exact stack overflow response haha